spec/services/shopify_refunder_spec.rb

```diff
@@ -0,0 +1,109 @@
+
+describe ShopifyRefunder do
+  let(:transaction_id) { '0xDEADBEEF' }
+  let(:options) { { reason: 'Actual reason', order_id: transaction_id } }
+  let(:transaction_amount) { 1 }
+  let(:credited_money_in_cents) { 100 }
+  let(:transaction) { double(:transaction, amount: transaction_amount, id: transaction_id) }
+
+  before do
+    allow(ShopifyAPI::Transaction).to receive(:find).and_return(transaction)
+  end
+
+  context '.initialize' do
+    subject { described_class.new(credited_money_in_cents, transaction_id, options) }
+
+    it 'successfully does it\'s thing' do
+      expect(subject).to be_truthy
+    end
+  end
+
+  context '.perform' do
+    let(:pos_refund) { double('refund', errors: []) }
+
+    before do
+      allow(ShopifyAPI::Refund).to receive(:create).and_return(pos_refund)
+    end
+
+    context 'when the shopify transaction is not found' do
+      before do
+        allow(ShopifyAPI::Transaction).to receive(:find).and_return(nil)
+      end
+
+      subject { described_class.new(credited_money_in_cents, transaction_id, options) }
+
+      it 'throws an error' do
+        cause = ->{ subject.perform }
+        expect(&cause).to raise_error(ActiveMerchant::Billing::ShopifyGateway::TransactionNotFoundError)
+      end
+    end
+
+    context 'with a full refund' do
+      subject { described_class.new(credited_money_in_cents, transaction_id, options) }
+
+      it 'performs a refund on shopify' do
+        expect(ShopifyAPI::Refund).to receive(:create)
+        subject.perform
+      end
+
+      it 'returns an ActiveMerchant response' do
+        result = subject.perform
+        expect(result).to be_a(ActiveMerchant::Billing::Response)
+      end
+
+      it 'returns an ActiveMerchant succesfull response' do
+        result = subject.perform
+        expect(result.success?).to be_truthy
+      end
+    end
+
+    context 'with a partial refund' do
+      let(:transaction_amount) { 2 }
+      let(:credited_money_in_cents) { 100 }
+
+      subject { described_class.new(credited_money_in_cents, transaction_id, options) }
+
+      it 'performs a refund on shopify' do
+        expect(ShopifyAPI::Refund).to receive(:create)
+        subject.perform
+      end
+
+      it 'returns an ActiveMerchant response' do
+        result = subject.perform
+        expect(result).to be_a(ActiveMerchant::Billing::Response)
+      end
+
+      it 'returns an ActiveMerchant succesfull response' do
+        result = subject.perform
+        expect(result.success?).to be_truthy
+      end
+    end
+
+    context 'with a credited amount bigger than the transaction' do
+      let(:transaction_amount) { 1 }
+      let(:credited_money_in_cents) { 200 }
+
+      subject { described_class.new(credited_money_in_cents, transaction_id, options) }
+
+      it 'throws an error' do
+        cause = ->{ subject.perform }
+        expect(&cause).to raise_error(ActiveMerchant::Billing::ShopifyGateway::CreditedAmountBiggerThanTransaction)
```

Comment:
I would suggest either a common namespace module (e.g. `Spree::Retail::Shopify`) or a superclass from which both inherit. The first option makes more sense to me, because I don't think inheritance makes sense. What do you think?